src/Raven.Server/Documents/Indexes/Index.cs

```diff
@@ -1399,24 +1401,26 @@ protected void ExecuteIndexing()
                         while (true)
                         {
 
-                            WaitHandle.WaitAny(new [] {_mre.WaitHandle, _rollingEvent.WaitHandle, _indexingProcessCancellationTokenSource.Token.WaitHandle});
+                            var result = WaitHandle.WaitAny(new [] {_mre.WaitHandle, _rollingEvent.WaitHandle, _indexingProcessCancellationTokenSource.Token.WaitHandle});
                             _indexingProcessCancellationTokenSource.Token.ThrowIfCancellationRequested();
                             
                             if (_indexDisabled)
                                 return;
-
-                            _rollingEvent.Reset();
+                            
+                            _mre.Reset();
 
                             var replaceStatus = ReplaceIfNeeded(batchCompleted: false, didWork: false);
 
                             if (replaceStatus == ReplaceStatus.Succeeded)
                                 return;
 
                             if (replaceStatus == ReplaceStatus.NotNeeded)
-                                break;
+                            {
+                                if (result != 0)
+                                    break;
+                            }
 
                             Thread.Sleep(500); // replace will be re-tried
```

Comment:
Why do we sleep here? Shouldn't we be doing all the waiting on the `WaitAny` ?

Reply:
to avoid a tight loop in case one of the events is set

Reply:
The typical way this is handled is that we'll either reset them (or if we won't need to listen to them any longer, not wait on them.

Reply:
This is important, since that can delay closing a db, for example.

Reply:
okay, will replace it with `_indexingProcessCancellationTokenSource.Token.WaitHandle.WaitOne(TimeSpan.FromMilliseconds(500));`